lib/streamlit/connections/base_connection.py

```diff
@@ -12,15 +12,67 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from abc import ABC
-from typing import Generic, TypeVar
+from abc import ABC, abstractmethod
+from typing import Any, Generic, Mapping, Optional, TypeVar
+
+from streamlit.runtime.secrets import AttrDict, secrets_singleton
 
 T = TypeVar("T")
 
 
 class BaseConnection(ABC, Generic[T]):
-    """TODO(vdonato): Implement... this entire class.
+    """TODO(vdonato): docstrings for this class and all public methods."""
+
+    def __init__(self, connection_name: str = "default", **kwargs) -> None:
+        self._connection_name = connection_name
+        self._kwargs = kwargs
+
+        self._raw_instance: Optional[T] = self.connect(**kwargs)
+
+    def _repr_html_(self) -> str:
+        # TODO(vdonato): Change this to whatever we actually want the default to be.
+        return f"Hi, I am a {self.default_connection_name()} connection!"
+
+    # Methods with default implementations that we don't expect subclasses to want or
+    # need to overwrite.
+    def get_secrets(self) -> Mapping[str, Any]:
```

Comment:
nit: I assume this is supposed to be part of the public API or? We probably could also add a `@gather_metrics("connection:get_secrets")` here as well.

Reply:
It's part of the public API but is intended to be used by connection authors rather than end-users (it's useful for implementing a `connect` function within a concrete connector), so I'm not sure if we'll want metrics on this since

CC @blackary in case you have an opinion on how useful it'd be to collect data on this